File: OrchidCore/src/main/java/com/eden/orchid/api/resources/resource/ResourceWrapper.java
package com.eden.orchid.api.resources.resource;

import com.eden.common.json.JSONElement;
import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.theme.pages.OrchidReference;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.InputStream;

/**
 * A Resource type that wraps another resource, optionally applying a transformation along the way.
 */
public class ResourceWrapper extends FreeableResource {

    private final OrchidResource resource;

    public ResourceWrapper(OrchidResource resource) {
        super(resource.getReference());
        this.resource = resource;
    }

    @Override
    public OrchidReference getReference() {
        return resource.getReference();
    }

    @Override
    public String getContent() {
        return resource.getContent();
    }

    @Override
    public JSONElement getEmbeddedData() {
        return resource.getEmbeddedData();
    }

    @Override
    public boolean shouldPrecompile() {
        return resource.shouldPrecompile();
    }

    @Override
    public boolean shouldRender() {
        return resource.shouldRender();
    }

    @Override
    @NotNull
    public InputStream getContentStream() {
        return resource.getContentStream();
    }

    @Override
    public String compileContent(Object data) {
        return resource.compileContent(data);
    }

    @Override
    public String getPrecompilerExtension() {
        return resource.getPrecompilerExtension();
    }

    @Override
    public boolean canUpdate() {
        return resource.canUpdate();
    }

    @Override
    public boolean canDelete() {
        return resource.canDelete();
    }

    @Override
    public void update(InputStream newContent) throws IOException {
        resource.update(newContent);
    }

    @Override
    public void delete() throws IOException {
        resource.delete();
    }

    @Override
    public void free() {
        resource.free();
    }

    @Override
    public String toString() {
        return resource.toString();
    }

    @Override
    public int hashCode() {
        return resource.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        return resource.equals(obj);
    }

}

File: OrchidCore/src/main/kotlin/com/eden/orchid/api/resources/resource/ResourceWrapper.kt
package com.eden.orchid.api.resources.resource

import com.eden.common.json.JSONElement
import com.eden.orchid.api.theme.pages.OrchidReference
import java.io.IOException
import java.io.InputStream

/**
 * A Resource type that wraps another resource, optionally applying a transformation along the way.
 */
abstract class ResourceWrapper(private val resource: OrchidResource) : OrchidResource(resource.reference) {

    override fun getContentStream(): InputStream {
        return resource.getContentStream()
    }

    override fun shouldPrecompile(): Boolean {
        return resource.shouldPrecompile()
    }

    override fun shouldRender(): Boolean {
        return resource.shouldRender()
    }

    override fun compileContent(data: Any?): String {
        return resource.compileContent(data)
    }

    override val precompilerExtension: String
        get() = resource.precompilerExtension

    override fun canUpdate(): Boolean {
        return resource.canUpdate()
    }

    override fun canDelete(): Boolean {
        return resource.canDelete()
    }

    override fun update(newContent: InputStream) {
        resource.update(newContent)
    }

    override fun delete() {
        resource.delete()
    }

    override fun toString(): String {
        return resource.toString()
    }

    override val content: String
        get() = resource.content

    override val embeddedData: JSONElement
        get() = resource.embeddedData

    override fun free() {
        resource.free()
    }

    override fun equals(other: Any?): Boolean {
        return resource.equals(other)
    }

    override fun hashCode(): Int {
        return resource.hashCode()
    }
}
